test/functional/p2p_dos_header_tree.py

```diff
@@ -62,7 +63,7 @@ def run_test(self):
 
         self.log.info("Feed all fork headers (succeeds without checkpoint)")
         # On node 0 it succeeds because checkpoints are disabled
-        self.restart_node(0, extra_args=['-nocheckpoints'])
+        self.restart_node(0, extra_args=['-nocheckpoints', "-minimumchainwork=0x0"])
```

Comment:
5691c174c5cb2f9947e6cc2db8d32b4b4715e47d: maybe add additional tests where you set `minimumchainwork` to the chainwork at the original checkpoint (`0x0000000000000000000000000000000000000000000000000000022302230223`).

```python

        self.log.info("Feed all fork headers (fails due to minimum chain work)")
        # Disable checkpoint, but require its chainwork as the minimum.
        self.restart_node(0, extra_args=['-nocheckpoints', "-minimumchainwork=0x0000000000000000000000000000000000000000000000000000022302230223"])
        peer_minimumchainwork = self.nodes[0].add_p2p_connection(P2PInterface())
        peer_minimumchainwork.send_and_ping(msg_headers(self.headers_fork))
        with self.nodes[0].assert_debug_log(['Ignoring low-work chain']):
            peer_minimumchainwork.send_message(msg_headers(self.headers_fork))
            # We don't disconnect
```

Note that if you also set the minimum chainwork for node 0 for the entire test, it will stall at "Feed all fork headers (fails due to checkpoint)", because we no longer disconnect.

Not sure if the above covers more ground than the new test file introduced in 7be81ce6257b1940d633b1311432c09fbbe09a0b, but it might still be useful to illustrate the behavior change.

Reply:
Do we need to rethink our headers sync timeout and/or disconnect logic? What happens when the first peer we try to do a headers sync with sends a low work chain? Since we don't disconnect, is there another mechanism that causes us to try with another peer, or do we wait for a timeout or new block announcement?

Reply:
@sjors Have you seen #25720?

Reply:
I have, but during initial sync it might be a while until the first INV arrives.

Reply:
This comment goes into more detail: https://github.com/bitcoin/bitcoin/pull/25720/files#r936707629

But still, if the first peer gives us low PoW headers - and we don't disconnect from it as we do know when it doesn't contain a checkpoint - that would cause us to do nothing for quite a while. At least IIUC.

Reply:
Oh wait, we don't have to disconnect, as long as `nSyncStarted` is set to 0 we'll try another peer.

Reply:
@Sjors Being in IBD doesn't stop other peers from announcing new blocks to us, IIRC.

Reply:
By "new blocks" do you mean blocks they know we don't have yet, or blocks they just received? Because the latter is not very frequent.

Reply:
Actually new blocks, once every 10 minutes. That's the current protection in practice against peers stalling headers sync forever by being very slow - once a new block is found on the network, we effectively start headers sync with all peers that announce it to us (unintentionally). #25720 changes it to just one new peer every block.

Reply:
So currently peers can stall us by sending valid headers slowly (ensuring the first checkpoint is not reached within ~10 minutes). After this PR they can stall us regardless of the how fast the headers are sent. But just sending headers slowly is simply enough to implement, so I suppose this PR does not make that problem worse.

Reply:
> So currently peers can stall us by sending valid headers slowly (ensuring the first checkpoint is not reached within ~10 minutes). After this PR they can stall us regardless of the how fast the headers are sent. But just sending headers slowly is simple enough to implement, so I suppose this PR does not make that problem worse.

This is not my understanding. Peers that we initially choose for headers sync currently have a time window of ~20 minutes (15min `HEADERS_DOWNLOAD_TIMEOUT_BASE` + a component based on the expected number of headers). If we don't have an up-to-date best header by then (no matter if we even received that from them or another peer), then they get disconnected. This logic is unaffected by this PR, although peers now have to serve the headers chain twice within these 20 minutes, so they need to be twice as fast.
Maybe it could make sense to add some criteria based on actual progress to this - if a peer has failed to serve us a single header for 10 minutes, does it really make sense to give them another 10 minutes?

The fact that we may also pick additional peers to sync headers from whenever a new block is found (10 minutes on average, but obviously very fluctuating) is independent from that, and also independent from the behavior of a potentially stalling peer. It could sometimes happen a few seconds after the headerssync start if a lucky block is mined then.

Reply:
That's still different from the case where a peer feeds us a set of headers that doesn't include a checkpoint. In that case we immediately disconnect and move on the next peer. That behavior goes away with this PR (though it's probably not a big deal).

Reply:
> Do we need to rethink our headers sync timeout and/or disconnect logic? What happens when the first peer we try to do a headers sync with sends a low work chain? Since we don't disconnect, is there another mechanism that causes us to try with another peer, or do we wait for a timeout or new block announcement? (trying to figure this out from just reading net_processing.cpp is daunting...)

> So currently peers can stall us by sending valid headers slowly (ensuring the first checkpoint is not reached within ~10 minutes). After this PR they can stall us regardless of the how fast the headers are sent. But just sending headers slowly is simple enough to implement, so I suppose this PR does not make that problem worse.

Note that ever since #5927, checkpoints only affect block/block header validation AFTER we have downloaded a headers chain that includes the checkpointed hashes.  So when a new node is starting up for the first time, if its initial headers sync peer serves a bogus chain, it would still be validated and accepted (even if it includes blocks at heights that have different block hashes checkpointed at those heights), because the main chain with the checkpointed block hashes wouldn't have been downloaded yet.

But that distinction is not so important, because for the purposes of (1) determining when the initial headers sync peer is bad/broken, or (2) determining whether a new node starting up can be flooded with low difficulty headers, an adversary has other options to cause us the same problems.  In the case of issue (1), as @mzumsande explained, for bandwidth reasons we give our initial headers sync peer roughly 20 minutes before we'll disconnect them if we haven't yet seen a chain that looks close to caught up.  So a peer looking to stall us can just give us _no headers at all_ (rather than give us a bogus headers chain) and can still stall us for the same amount of time, under that check.  In the case of issue (2), even if block hashes were being checked at the checkpointed heights prior to downloading a chain with those hashes, an adversary looking to waste our memory could just give us a lots of chains that start at genesis and build up to the first checkpoint, and then restart from genesis again with a new chain.

I believe this PR completely solves issue (2).

Regarding issue (1): for a very long time, our protection against a stalling peer has been that we will initiate headers sync (via a `getheaders` message) with any peer sending us a block INV.  This is overkill and made somewhat more bandwidth-wasteful after this PR, so #25720 tries to strike a better balance between robustness in the face of stalling peers, and bandwidth usage.  However I believe exactly how we strike that balance is orthogonal to the behavior introduced here, so it makes more sense to discuss strategies for when we start to sync headers with a peer in that PR, rather than here.